In [1]:
import pandas as pd
from pathlib import Path

# Chargement de données

In [5]:
# Chemin d'accès aux données brutes

RAW_DATA_DIR = Path("../data/raw/")
# Charger les datasets

files = {
    "idmc": "data_idmc_depuis_2000.csv",
    "solutions": "data_solutions_depuis_2000.csv",
    "decisions": "decisions_asile_depuis_2000.csv",
    "demandes": "demandes_asile_depuis_2000.csv",
    "demographie": "demographie_depuis_2000.csv",
    "pays": "countries.csv"
}


dfs = {
    name: pd.read_csv(RAW_DATA_DIR / filename)
    for name, filename in files.items()
}

for name, df in dfs.items():
    print(f"{name:15} : {df.shape[0]:,} lignes × {df.shape[1]} colonnes")

idmc            : 934 lignes × 10 colonnes
solutions       : 21,258 lignes × 13 colonnes
decisions       : 113,929 lignes × 17 colonnes
demandes        : 120,597 lignes × 14 colonnes
demographie     : 116,781 lignes × 24 colonnes
pays            : 232 lignes × 16 colonnes


## 1. Structure et completude

In [6]:
# Vérification de la structure

for name, df in dfs.items():
    print("\n", "=" * 70)
    print(name.upper())
    print("=" * 70)

    print("Shape :", df.shape)
    print("\nColonnes :")
    print(df.columns.tolist())

    print("\nTypes :")
    print(df.dtypes)

    print("\nAperçu :")
    display(df.head())


IDMC
Shape : (934, 10)

Colonnes :
['year', 'coo_id', 'coo_name', 'coo', 'coo_iso', 'coa_id', 'coa_name', 'coa', 'coa_iso', 'total']

Types :
year         int64
coo_id       int64
coo_name    object
coo         object
coo_iso     object
coa_id       int64
coa_name    object
coa         object
coa_iso     object
total        int64
dtype: object

Aperçu :


,year,coo_id,coo_name,coo,coo_iso,coa_id,coa_name,coa,coa_iso,total
0,1990,262,Unknown,UKN,UNK,262,Unknown,UKN,UNK,21300000
1,1991,262,Unknown,UKN,UNK,262,Unknown,UKN,UNK,22500000
2,1992,262,Unknown,UKN,UNK,262,Unknown,UKN,UNK,25000000
3,1993,262,Unknown,UKN,UNK,262,Unknown,UKN,UNK,27500000
4,1994,262,Unknown,UKN,UNK,262,Unknown,UKN,UNK,28000000



SOLUTIONS
Shape : (21258, 13)

Colonnes :
['year', 'coo_id', 'coo_name', 'coo', 'coo_iso', 'coa_id', 'coa_name', 'coa', 'coa_iso', 'returned_refugees', 'resettlement', 'naturalisation', 'returned_idps']

Types :
year                  int64
coo_id                int64
coo_name             object
coo                  object
coo_iso              object
coa_id                int64
coa_name             object
coa                  object
coa_iso              object
returned_refugees    object
resettlement         object
naturalisation       object
returned_idps        object
dtype: object

Aperçu :


,year,coo_id,coo_name,coo,coo_iso,coa_id,coa_name,coa,coa_iso,returned_refugees,resettlement,naturalisation,returned_idps
0,1959,3,Albania,ALB,ALB,33,Canada,CAN,CAN,-,17,-,-
1,1959,8,Egypt,ARE,EGY,33,Canada,CAN,CAN,-,5,-,-
2,1959,12,Austria,AUS,AUT,33,Canada,CAN,CAN,-,17,-,-
3,1959,30,Bulgaria,BUL,BGR,33,Canada,CAN,CAN,-,35,-,-
4,1959,49,Czechia,CZE,CZE,33,Canada,CAN,CAN,-,36,-,-



DECISIONS
Shape : (113929, 17)

Colonnes :
['year', 'coo_id', 'coo_name', 'coo', 'coo_iso', 'coa_id', 'coa_name', 'coa', 'coa_iso', 'procedure_type', 'dec_level', 'dec_pc', 'dec_recognized', 'dec_other', 'dec_rejected', 'dec_closed', 'dec_total']

Types :
year               int64
coo_id             int64
coo_name          object
coo               object
coo_iso           object
coa_id             int64
coa_name          object
coa               object
coa_iso           object
procedure_type    object
dec_level         object
dec_pc            object
dec_recognized     int64
dec_other          int64
dec_rejected       int64
dec_closed         int64
dec_total          int64
dtype: object

Aperçu :


,year,coo_id,coo_name,coo,coo_iso,coa_id,coa_name,coa,coa_iso,procedure_type,dec_level,dec_pc,dec_recognized,dec_other,dec_rejected,dec_closed,dec_total
0,2000,2,Afghanistan,AFG,AFG,11,Australia,AUL,AUS,G,AR,C,38,0,25,0,63
1,2000,3,Albania,ALB,ALB,11,Australia,AUL,AUS,G,AR,C,5,0,24,0,29
2,2000,4,Algeria,ALG,DZA,11,Australia,AUL,AUS,G,AR,C,5,0,17,0,22
3,2000,8,Egypt,ARE,EGY,11,Australia,AUL,AUS,G,AR,C,21,0,50,5,76
4,2000,10,Armenia,ARM,ARM,11,Australia,AUL,AUS,G,AR,C,0,0,5,0,5



DEMANDES
Shape : (120597, 14)

Colonnes :
['year', 'coo_id', 'coo_name', 'coo', 'coo_iso', 'coa_id', 'coa_name', 'coa', 'coa_iso', 'procedure_type', 'app_type', 'dec_level', 'app_pc', 'applied']

Types :
year               int64
coo_id             int64
coo_name          object
coo               object
coo_iso           object
coa_id             int64
coa_name          object
coa               object
coa_iso           object
procedure_type    object
app_type          object
dec_level         object
app_pc            object
applied            int64
dtype: object

Aperçu :


,year,coo_id,coo_name,coo,coo_iso,coa_id,coa_name,coa,coa_iso,procedure_type,app_type,dec_level,app_pc,applied
0,2006,2,Afghanistan,AFG,AFG,11,Australia,AUL,AUS,G,A,AR,C,14
1,2006,3,Albania,ALB,ALB,11,Australia,AUL,AUS,G,A,AR,C,21
2,2006,4,Algeria,ALG,DZA,11,Australia,AUL,AUS,G,A,AR,C,5
3,2006,8,Egypt,ARE,EGY,11,Australia,AUL,AUS,G,A,AR,C,38
4,2006,14,Bahrain,BAH,BHR,11,Australia,AUL,AUS,G,A,AR,C,11



DEMOGRAPHIE
Shape : (116781, 24)

Colonnes :
['year', 'coo_id', 'coo_name', 'coo', 'coo_iso', 'coa_id', 'coa_name', 'coa', 'coa_iso', 'f_0_4', 'f_5_11', 'f_12_17', 'f_18_59', 'f_60', 'f_other', 'f_total', 'm_0_4', 'm_5_11', 'm_12_17', 'm_18_59', 'm_60', 'm_other', 'm_total', 'total']

Types :
year         int64
coo_id       int64
coo_name    object
coo         object
coo_iso     object
coa_id       int64
coa_name    object
coa         object
coa_iso     object
f_0_4        int64
f_5_11       int64
f_12_17      int64
f_18_59      int64
f_60         int64
f_other      int64
f_total      int64
m_0_4        int64
m_5_11       int64
m_12_17      int64
m_18_59      int64
m_60         int64
m_other      int64
m_total      int64
total        int64
dtype: object

Aperçu :


,year,coo_id,coo_name,coo,coo_iso,coa_id,coa_name,coa,coa_iso,f_0_4,...,f_other,f_total,m_0_4,m_5_11,m_12_17,m_18_59,m_60,m_other,m_total,total
0,2001,2,Afghanistan,AFG,AFG,2,Afghanistan,AFG,AFG,0,...,0,0,0,0,0,0,0,0,0,1226092
1,2001,91,Iran (Islamic Rep. of),IRN,IRN,2,Afghanistan,AFG,AFG,0,...,0,0,0,0,0,5,0,0,5,5
2,2001,92,Iraq,IRQ,IRQ,2,Afghanistan,AFG,AFG,0,...,0,5,0,0,0,0,0,0,0,5
3,2001,92,Iraq,IRQ,IRQ,3,Albania,ALB,ALB,0,...,0,0,0,0,0,7,0,0,7,7
4,2001,120,North Macedonia,MCD,MKD,3,Albania,ALB,ALB,0,...,0,14,0,0,0,21,0,0,21,35



PAYS
Shape : (232, 16)

Colonnes :
['0', 'id', 'code', 'iso', 'iso2', 'name', 'nameOrigin', 'nameLong', 'nameShort', 'nameFormal', 'nationality', 'majorArea', 'region', 'nameFr', 'majorAreaFr', 'regionFr']

Types :
0               int64
id              int64
code           object
iso            object
iso2           object
name           object
nameOrigin     object
nameLong       object
nameShort      object
nameFormal     object
nationality    object
majorArea      object
region         object
nameFr         object
majorAreaFr    object
regionFr       object
dtype: object

Aperçu :


,0,id,code,iso,iso2,name,nameOrigin,nameLong,nameShort,nameFormal,nationality,majorArea,region,nameFr,majorAreaFr,regionFr
0,1,2,AFG,AFG,AF,Afghanistan,Afghanistan,Afghanistan,Afghanistan,the Islamic State of Afghanistan,Afghan,Asia,Southern Asia,Afghanistan,Asie,Sud de l'Asie centrale
1,1,3,ALB,ALB,AL,Albania,Albania,Albania,Albania,the Republic of Albania,Albanian,Europe,Southern Europe,Albanie,Europe,Europe australe
2,1,4,ALG,DZA,DZ,Algeria,Algeria,Algeria,Algeria,the People's Democratic Republic of Algeria,Algerian,Africa,Northern Africa,Alg,Afrique,Afrique septentrionale
3,1,5,AND,AND,AD,Andorra,Andorra,Andorra,Andorra,the Principality of Andorra,Andorran,Europe,Southern Europe,Andorre,Europe,Europe australe
4,1,6,ANG,AGO,AO,Angola,Angola,Angola,Angola,the Republic of Angola,Angolan,Africa,Middle Africa,Angola,Afrique,Afrique australe


Pour chaque colonne, on calcul le nombre et le pourcentage de valeurs manquantes.

In [7]:
def audit_missing(df):
    audit = pd.DataFrame({
        "nb_lignes": len(df),
        "nb_manquants": df.isna().sum(),
        "taux_manquants_pct": df.isna().mean() * 100
    })

    return audit.sort_values(
        "taux_manquants_pct",
        ascending=False
    )

for name, df in dfs.items():
    print(f"\n### {name.upper()}")
    display(audit_missing(df))


### IDMC


,nb_lignes,nb_manquants,taux_manquants_pct
year,934,0,0.0
coo_id,934,0,0.0
coo_name,934,0,0.0
coo,934,0,0.0
coo_iso,934,0,0.0
coa_id,934,0,0.0
coa_name,934,0,0.0
coa,934,0,0.0
coa_iso,934,0,0.0
total,934,0,0.0



### SOLUTIONS


,nb_lignes,nb_manquants,taux_manquants_pct
year,21258,0,0.0
coo_id,21258,0,0.0
coo_name,21258,0,0.0
coo,21258,0,0.0
coo_iso,21258,0,0.0
coa_id,21258,0,0.0
coa_name,21258,0,0.0
coa,21258,0,0.0
coa_iso,21258,0,0.0
returned_refugees,21258,0,0.0



### DECISIONS


,nb_lignes,nb_manquants,taux_manquants_pct
dec_level,113929,2139,1.877485
year,113929,0,0.000000
procedure_type,113929,0,0.000000
dec_closed,113929,0,0.000000
dec_rejected,113929,0,0.000000
dec_other,113929,0,0.000000
dec_recognized,113929,0,0.000000
dec_pc,113929,0,0.000000
coa_iso,113929,0,0.000000
coo_id,113929,0,0.000000



### DEMANDES


,nb_lignes,nb_manquants,taux_manquants_pct
dec_level,120597,2180,1.807673
app_type,120597,358,0.296856
year,120597,0,0.000000
coo_id,120597,0,0.000000
coo_name,120597,0,0.000000
coo,120597,0,0.000000
coo_iso,120597,0,0.000000
coa_id,120597,0,0.000000
coa_name,120597,0,0.000000
coa,120597,0,0.000000



### DEMOGRAPHIE


,nb_lignes,nb_manquants,taux_manquants_pct
year,116781,0,0.0
coo_id,116781,0,0.0
m_total,116781,0,0.0
m_other,116781,0,0.0
m_60,116781,0,0.0
m_18_59,116781,0,0.0
m_12_17,116781,0,0.0
m_5_11,116781,0,0.0
m_0_4,116781,0,0.0
f_total,116781,0,0.0



### PAYS


,nb_lignes,nb_manquants,taux_manquants_pct
nationality,232,39,16.810345
nameFormal,232,32,13.793103
nameLong,232,17,7.327586
majorAreaFr,232,17,7.327586
regionFr,232,17,7.327586
nameFr,232,14,6.034483
iso2,232,8,3.448276
majorArea,232,5,2.155172
region,232,5,2.155172
nameOrigin,232,4,1.724138


## 2. Unicité

In [8]:
### Unicité : recherche des doublons stricts

for name, df in dfs.items():

    n_dup = df.duplicated().sum()
    pct_dup = 100 * n_dup / len(df) if len(df) else 0

    print(
        f"{name:15} "
        f"{n_dup:,} doublons stricts "
        f"({pct_dup:.2f} %)"
    )

idmc            0 doublons stricts (0.00 %)
solutions       0 doublons stricts (0.00 %)
decisions       0 doublons stricts (0.00 %)
demandes        0 doublons stricts (0.00 %)
demographie     0 doublons stricts (0.00 %)
pays            0 doublons stricts (0.00 %)


In [9]:
doublons = dfs["idmc"][
    dfs["idmc"].duplicated(keep=False)
]

display(doublons)

,year,coo_id,coo_name,coo,coo_iso,coa_id,coa_name,coa,coa_iso,total


In [10]:
doublons = dfs["solutions"][
    dfs["solutions"].duplicated(keep=False)
]

display(doublons)

,year,coo_id,coo_name,coo,coo_iso,coa_id,coa_name,coa,coa_iso,returned_refugees,resettlement,naturalisation,returned_idps


In [11]:
doublons = dfs["decisions"][
    dfs["decisions"].duplicated(keep=False)
]

display(doublons)

,year,coo_id,coo_name,coo,coo_iso,coa_id,coa_name,coa,coa_iso,procedure_type,dec_level,dec_pc,dec_recognized,dec_other,dec_rejected,dec_closed,dec_total


In [12]:
doublons = dfs["demandes"][
    dfs["demandes"].duplicated(keep=False)
]

display(doublons)

,year,coo_id,coo_name,coo,coo_iso,coa_id,coa_name,coa,coa_iso,procedure_type,app_type,dec_level,app_pc,applied


In [13]:
doublons = dfs["demographie"][
    dfs["demographie"].duplicated(keep=False)
]

display(doublons)

,year,coo_id,coo_name,coo,coo_iso,coa_id,coa_name,coa,coa_iso,f_0_4,...,f_other,f_total,m_0_4,m_5_11,m_12_17,m_18_59,m_60,m_other,m_total,total


In [14]:
doublons = dfs["pays"][
    dfs["pays"].duplicated(keep=False)
]
display(doublons)

,0,id,code,iso,iso2,name,nameOrigin,nameLong,nameShort,nameFormal,nationality,majorArea,region,nameFr,majorAreaFr,regionFr


### Recherche des doublons métier
*Deux lignes peuvent être différentes techniquement mais représenter potentiellement le même objet métier.*

In [15]:
# Demandes d'asile

key_demandes = [
    "year",
    "coo_id",
    "coa_id",
    "procedure_type",
    "app_type",
    "dec_level",
    "app_pc"
]

key = [c for c in key_demandes if c in dfs["demandes"].columns]

duplicates_business = (
    dfs["demandes"]
    .loc[lambda x: x.duplicated(key, keep=False)]
    .sort_values(key)
)

display(duplicates_business)

,year,coo_id,coo_name,coo,coo_iso,coa_id,coa_name,coa,coa_iso,procedure_type,app_type,dec_level,app_pc,applied


In [16]:
# Décision d'asile
key_decisions = [
    "year",
    "coo_id",
    "coa_id",
    "procedure_type",
    "dec_level",
    "dec_pc"
]

key = [c for c in key_decisions if c in dfs["decisions"].columns]

duplicates_business = (
    dfs["decisions"]
    .loc[lambda x: x.duplicated(key, keep=False)]
    .sort_values(key)
)

display(duplicates_business)

,year,coo_id,coo_name,coo,coo_iso,coa_id,coa_name,coa,coa_iso,procedure_type,dec_level,dec_pc,dec_recognized,dec_other,dec_rejected,dec_closed,dec_total


In [17]:
# IDMC
key_idmc = [
    "year",
    "coo_id",
    "coa_id"
]

key = [c for c in key_idmc if c in dfs["idmc"].columns]

duplicates_business = (
    dfs["idmc"]
    .loc[lambda x: x.duplicated(key, keep=False)]
    .sort_values(key)
)

display(duplicates_business)

,year,coo_id,coo_name,coo,coo_iso,coa_id,coa_name,coa,coa_iso,total


In [18]:
# Solutions
key_solutions = [
    "year",
    "coo_id",
    "coa_id"
]

key = [c for c in key_solutions if c in dfs["solutions"].columns]

duplicates_business = (
    dfs["solutions"]
    .loc[lambda x: x.duplicated(key, keep=False)]
    .sort_values(key)
)

display(duplicates_business)

,year,coo_id,coo_name,coo,coo_iso,coa_id,coa_name,coa,coa_iso,returned_refugees,resettlement,naturalisation,returned_idps


In [19]:
# Demographie
key_demographie = [
    "year",
    "coo_id",
    "coa_id"
]
key = [c for c in key_demographie if c in dfs["demographie"].columns]

duplicates_business = (
    dfs["demographie"]
    .loc[lambda x: x.duplicated(key, keep=False)]
    .sort_values(key)
)

display(duplicates_business)

,year,coo_id,coo_name,coo,coo_iso,coa_id,coa_name,coa,coa_iso,f_0_4,...,f_other,f_total,m_0_4,m_5_11,m_12_17,m_18_59,m_60,m_other,m_total,total


In [20]:
# Referentiel Pays
key = "id"

duplicates_business = (
    dfs["pays"]
    .loc[lambda x: x.duplicated(key, keep=False)]
    .sort_values(key)
)

display(duplicates_business)

,0,id,code,iso,iso2,name,nameOrigin,nameLong,nameShort,nameFormal,nationality,majorArea,region,nameFr,majorAreaFr,regionFr


## 3. Accuracy

In [22]:
### Cohérence géographique de l’IDMC
df = dfs["idmc"].copy()

coo = df["coo_iso"].astype("string").str.strip().str.upper()
coa = df["coa_iso"].astype("string").str.strip().str.upper()

same_country = coo == coa

print("Même origine/destination :", same_country.sum())
print("Différents :", (~same_country).sum())

Même origine/destination : 934
Différents : 0


In [23]:
display(
    df.loc[
        ~same_country,
        [
            "year",
            "coo_id",
            "coo_name",
            "coo_iso",
            "coa_id",
            "coa_name",
            "coa_iso",
            "total"
        ]
    ].head(50)
)

,year,coo_id,coo_name,coo_iso,coa_id,coa_name,coa_iso,total


In [24]:
### Cohérence des décisions d’asile
df = dfs["decisions"].copy()

cols = [
    "dec_recognized",
    "dec_other",
    "dec_rejected",
    "dec_closed",
    "dec_total"
]

for col in cols:
    df[col] = pd.to_numeric(df[col], errors="coerce")

df["dec_sum"] = (
    df[
        [
            "dec_recognized",
            "dec_other",
            "dec_rejected",
            "dec_closed"
        ]
    ]
    .sum(axis=1, min_count=1)
)

df["ecart"] = df["dec_total"] - df["dec_sum"]

display(
    df[df["ecart"].fillna(0) != 0].head()
)

,year,coo_id,coo_name,coo,coo_iso,coa_id,coa_name,coa,coa_iso,procedure_type,dec_level,dec_pc,dec_recognized,dec_other,dec_rejected,dec_closed,dec_total,dec_sum,ecart
73,2000,10,Armenia,ARM,ARM,49,Czechia,CZE,CZE,G,AR,C,15,0,21,5,43,41,2
77,2000,30,Bulgaria,BUL,BGR,49,Czechia,CZE,CZE,G,AR,C,5,0,75,10,89,90,-1
78,2000,40,Congo,COB,COG,49,Czechia,CZE,CZE,G,AR,C,0,0,5,10,14,15,-1
100,2000,200,Ukraine,UKR,UKR,49,Czechia,CZE,CZE,G,AR,C,5,0,39,19,64,63,1
201,2001,262,Unknown,UKN,UNK,132,Malta,MTA,MLT,G,AR,C,0,0,0,0,5,0,5


In [25]:
### Cohérence démographique

df = dfs["demographie"].copy()

for col in ["f_total", "m_total", "total"]:
    df[col] = pd.to_numeric(df[col], errors="coerce")
df["total_calcule"] = (
    df["f_total"] + df["m_total"]
)

df["ecart"] = (
    df["total"] - df["total_calcule"]
)

print(
    "Lignes avec écart :",
    (df["ecart"].fillna(0) != 0).sum()
)

Lignes avec écart : 66185


## 4. Cohérence

In [29]:
# On cherche maintenant les différences susceptibles de casser les futures jointures.

text_cols = [
    "coo",
    "coo_name",
    "coo_iso",
    "coa",
    "coa_name",
    "coa_iso"
]

for name, df in dfs.items():

    for col in text_cols:

        if col in df.columns:

            nb_spaces = (
                df[col].dropna().astype(str)
                != df[col].dropna().astype(str).str.strip()
            ).sum()

            print(name, col, "espaces :", nb_spaces)

idmc coo espaces : 0
idmc coo_name espaces : 22
idmc coo_iso espaces : 0
idmc coa espaces : 0
idmc coa_name espaces : 22
idmc coa_iso espaces : 0
solutions coo espaces : 0
solutions coo_name espaces : 597
solutions coo_iso espaces : 0
solutions coa espaces : 0
solutions coa_name espaces : 138
solutions coa_iso espaces : 0
decisions coo espaces : 0
decisions coo_name espaces : 1166
decisions coo_iso espaces : 0
decisions coa espaces : 0
decisions coa_name espaces : 20
decisions coa_iso espaces : 0
demandes coo espaces : 0
demandes coo_name espaces : 1187
demandes coo_iso espaces : 0
demandes coa espaces : 0
demandes coa_name espaces : 20
demandes coa_iso espaces : 0
demographie coo espaces : 0
demographie coo_name espaces : 1282
demographie coo_iso espaces : 0
demographie coa espaces : 0
demographie coa_name espaces : 52
demographie coa_iso espaces : 0


In [30]:
## Validité des années

for name, df in dfs.items():

    # Exclure le df "pays"
    if name == "pays":
        continue

    year = pd.to_numeric(df["year"], errors="coerce")

    invalid = df[
        year.isna() |
        ~year.between(2000, 2025)
    ]

    print(
        name,
        "min =", year.min(),
        "max =", year.max(),
        "invalides =", len(invalid)
    )

idmc min = 1990 max = 2025 invalides = 10
solutions min = 1959 max = 2025 invalides = 5443
decisions min = 2000 max = 2025 invalides = 0
demandes min = 2000 max = 2025 invalides = 0
demographie min = 2001 max = 2025 invalides = 0


In [31]:
## Validité des codes pays
## C'est essentiel puisque coo_iso et coa_iso vont servir aux analyses géographiques et aux jointures.

for name, df in dfs.items():

    for col in ["coo_iso", "coa_iso"]:

        if col in df.columns:

            print(
                name,
                col,
                df[col].nunique(dropna=False)
            )

            print(
                sorted(
                    df[col]
                    .dropna()
                    .astype(str)
                    .unique()
                )[:30]
            )

idmc coo_iso 84
['AB9', 'AFG', 'ARM', 'AZE', 'BDI', 'BEN', 'BFA', 'BGD', 'BIH', 'BOL', 'BRA', 'CAF', 'CIV', 'CMR', 'COD', 'COG', 'COL', 'COM', 'CYP', 'DZA', 'ECU', 'EGY', 'ERI', 'ETH', 'GEO', 'GHA', 'GMB', 'GTM', 'HND', 'HRV']
idmc coa_iso 84
['AB9', 'AFG', 'ARM', 'AZE', 'BDI', 'BEN', 'BFA', 'BGD', 'BIH', 'BOL', 'BRA', 'CAF', 'CIV', 'CMR', 'COD', 'COG', 'COL', 'COM', 'CYP', 'DZA', 'ECU', 'EGY', 'ERI', 'ETH', 'GEO', 'GHA', 'GMB', 'GTM', 'HND', 'HRV']
solutions coo_iso 189
['AFG', 'AGO', 'ALB', 'ARE', 'ARG', 'ARM', 'ATG', 'AUS', 'AUT', 'AZE', 'BDI', 'BEL', 'BEN', 'BFA', 'BGD', 'BGR', 'BHR', 'BHS', 'BIH', 'BLR', 'BLZ', 'BOL', 'BRA', 'BRB', 'BRN', 'BTN', 'BWA', 'CAF', 'CAN', 'CHE']
solutions coa_iso 171
['AFG', 'AGO', 'ALB', 'ARE', 'ARG', 'ARM', 'ATG', 'AUS', 'AUT', 'AZE', 'BDI', 'BEL', 'BEN', 'BFA', 'BGD', 'BGR', 'BHR', 'BHS', 'BIH', 'BLR', 'BLZ', 'BOL', 'BRA', 'BWA', 'CAF', 'CAN', 'CHE', 'CHL', 'CHN', 'CIV']
decisions coo_iso 211
['AFG', 'AGO', 'ALB', 'AND', 'ARE', 'ARG', 'ARM', 'ATG', '

In [32]:
pattern_iso3 = r"^[A-Z]{3}$"

for name, df in dfs.items():

    for col in ["coo_iso", "coa_iso"]:

        if col in df.columns:

            values = df[col].astype("string").str.strip().str.upper()

            invalid = (
                values.notna()
                & ~values.str.match(pattern_iso3, na=False)
            )

            print(name, col, "formats invalides :", invalid.sum())

idmc coo_iso formats invalides : 12
idmc coa_iso formats invalides : 12
solutions coo_iso formats invalides : 0
solutions coa_iso formats invalides : 0
decisions coo_iso formats invalides : 0
decisions coa_iso formats invalides : 0
demandes coo_iso formats invalides : 0
demandes coa_iso formats invalides : 0
demographie coo_iso formats invalides : 0
demographie coa_iso formats invalides : 0


In [33]:
# Vérification des formats invalides de IDMC

idmc = dfs["idmc"]

for col in ["coo_iso", "coa_iso"]:

    invalid_mask = (
        idmc[col].isna() |
        ~idmc[col]
            .astype(str)
            .str.strip()
            .str.upper()
            .str.fullmatch(r"[A-Z]{3}")
    )

    print(f"\n--- {col} ---")

    print(
        idmc.loc[invalid_mask, col]
            .value_counts(dropna=False)
    )


--- coo_iso ---
coo_iso
AB9    12
Name: count, dtype: int64

--- coa_iso ---
coa_iso
AB9    12
Name: count, dtype: int64


In [34]:
### Validité des mesures numériques

numeric_expected = {
    "idmc": ["total"],

    "solutions": [
        "returned_refugees",
        "resettlement",
        "naturalisation",
        "returned_idps"
    ],

    "decisions": [
        "dec_recognized",
        "dec_other",
        "dec_rejected",
        "dec_closed",
        "dec_total"
    ],

    "demandes": ["applied"],

    "demographie": [
        "f_0_4", "f_5_11", "f_12_", "f_18_59",
        "f_60", "f_other", "f_total",
        "m_0_4", "m_5_11", "m_12_17", "m_18_59",
        "m_60", "m_other", "m_total",
        "total"
    ]
}

In [35]:
for name, cols in numeric_expected.items():

    df = dfs[name]

    for col in cols:

        if col not in df.columns:
            continue

        converted = pd.to_numeric(df[col], errors="coerce")

        conversion_errors = (
            df[col].notna() &
            converted.isna()
        ).sum()

        negative = (converted < 0).sum()

        print(
            name,
            col,
            "non numériques =", conversion_errors,
            "| négatifs =", negative
        )

idmc total non numériques = 0 | négatifs = 0
solutions returned_refugees non numériques = 15161 | négatifs = 0
solutions resettlement non numériques = 11644 | négatifs = 0
solutions naturalisation non numériques = 13872 | négatifs = 0
solutions returned_idps non numériques = 20878 | négatifs = 0
decisions dec_recognized non numériques = 0 | négatifs = 0
decisions dec_other non numériques = 0 | négatifs = 0
decisions dec_rejected non numériques = 0 | négatifs = 0
decisions dec_closed non numériques = 0 | négatifs = 0
decisions dec_total non numériques = 0 | négatifs = 0
demandes applied non numériques = 0 | négatifs = 0
demographie f_0_4 non numériques = 0 | négatifs = 0
demographie f_5_11 non numériques = 0 | négatifs = 0
demographie f_18_59 non numériques = 0 | négatifs = 0
demographie f_60 non numériques = 0 | négatifs = 0
demographie f_other non numériques = 0 | négatifs = 0
demographie f_total non numériques = 0 | négatifs = 0
demographie m_0_4 non numériques = 0 | négatifs = 0
dem

In [36]:
# Identification des valeurs non numériques de df solutions

df = dfs["solutions"]

cols = [
    "returned_refugees",
    "resettlement",
    "naturalisation",
    "returned_idps"
]

for col in cols:

    converted = pd.to_numeric(df[col], errors="coerce")

    invalid_mask = (
        df[col].notna() &
        converted.isna()
    )

    print(f"\n--- {col} ---")
    print(f"Nombre d'anomalies : {invalid_mask.sum()}")

    print(
        df.loc[invalid_mask, col]
        .value_counts(dropna=False)
        .head(20)
    )


--- returned_refugees ---
Nombre d'anomalies : 15161
returned_refugees
-    15161
Name: count, dtype: int64

--- resettlement ---
Nombre d'anomalies : 11644
resettlement
-    11644
Name: count, dtype: int64

--- naturalisation ---
Nombre d'anomalies : 13872
naturalisation
-    13872
Name: count, dtype: int64

--- returned_idps ---
Nombre d'anomalies : 20878
returned_idps
-    20878
Name: count, dtype: int64


## 5. Timeliness

In [37]:
### Timeliness : couverture temporelle des sources

coverage = []

for name in [
    "idmc",
    "demandes",
    "decisions",
    "solutions",
    "demographie"
]:

    year = pd.to_numeric(
        dfs[name]["year"],
        errors="coerce"
    )

    coverage.append({
        "dataset": name,
        "min": year.min(),
        "max": year.max(),
        "nb_annees": year.nunique(),
        "nb_lignes": len(dfs[name])
    })

coverage = pd.DataFrame(coverage)

display(coverage)

,dataset,min,max,nb_annees,nb_lignes
0,idmc,1990,2025,36,934
1,demandes,2000,2025,26,120597
2,decisions,2000,2025,26,113929
3,solutions,1959,2025,67,21258
4,demographie,2001,2025,25,116781


In [38]:
### Vérification des écarts dans la colonne "year"

idmc = dfs["idmc"].copy()

idmc_year = (
    idmc
    .sort_values(["coo_id", "year"])
    .copy()
)

idmc_year["previous_year"] = (
    idmc_year
    .groupby("coo_id")["year"]
    .shift(1)
)

idmc_year["year_gap"] = (
    idmc_year["year"]
    - idmc_year["previous_year"]
)

# Compter la fréquence des écarts
display(
    idmc_year["year_gap"]
    .value_counts()
    .sort_index()
)

year_gap
1.0    835
2.0      6
3.0      2
4.0      3
5.0      1
6.0      1
7.0      1
9.0      1
Name: count, dtype: int64

La très grande majorité des observations successives sont espacées exactement d’un an, 
mais il existe quelques ruptures temporelles qu’il faut investiguer.


98,2 % des 850 transitions temporelles observées sont annuelles. 15 transitions (1,8 %) présentent une discontinuité supérieure à un an, avec un écart maximal de 9 ans. Ces cas doivent être investigués avant toute imputation ou exclusion.

In [39]:
### Recherche des coo_id et des années avec rupture

gaps = (
    idmc_year
    .loc[idmc_year["year_gap"] > 1,
         ["coo_id", "previous_year", "year", "year_gap"]]
    .sort_values(["year_gap", "coo_id"], ascending=[False, True])
)

display(gaps)

,coo_id,previous_year,year,year_gap
474,262,2008.0,2017,9.0
560,100,2012.0,2019,7.0
757,192,2016.0,2022,6.0
783,73,2018.0,2023,5.0
591,10,2016.0,2020,4.0
930,188,2021.0,2025,4.0
699,262,2017.0,2021,4.0
676,117,2018.0,2021,3.0
853,117,2021.0,2024,3.0
733,100,2020.0,2022,2.0


In [40]:
### L'historique complet des coo_id concernés

coo_with_gaps = (
    idmc_year
    .loc[idmc_year["year_gap"] > 1, "coo_id"]
    .unique()
)

display(
    idmc_year[
        idmc_year["coo_id"].isin(coo_with_gaps)
    ]
    .sort_values(["coo_id", "year"])
)

,year,coo_id,coo_name,coo,coo_iso,coa_id,coa_name,coa,coa_iso,total,previous_year,year_gap
20,2009,10,Armenia,ARM,ARM,10,Armenia,ARM,ARM,8400,NaN,NaN
66,2010,10,Armenia,ARM,ARM,10,Armenia,ARM,ARM,8400,2009.0,1.0
113,2011,10,Armenia,ARM,ARM,10,Armenia,ARM,ARM,8400,2010.0,1.0
162,2012,10,Armenia,ARM,ARM,10,Armenia,ARM,ARM,8400,2011.0,1.0
211,2013,10,Armenia,ARM,ARM,10,Armenia,ARM,ARM,8400,2012.0,1.0
...,...,...,...,...,...,...,...,...,...,...,...,...
17,2007,262,Unknown,UKN,UNK,262,Unknown,UKN,UNK,26000000,2006.0,1.0
18,2008,262,Unknown,UKN,UNK,262,Unknown,UKN,UNK,26000000,2007.0,1.0
474,2017,262,Unknown,UKN,UNK,262,Unknown,UKN,UNK,2,2008.0,9.0
699,2021,262,Unknown,UKN,UNK,262,Unknown,UKN,UNK,8,2017.0,4.0


Parmi 850 transitions observées entre deux enregistrements successifs d'un même pays d'origine, 835 (98,2 %) sont espacées d'un an et 15 (1,8 %) présentent un intervalle supérieur à un an, avec un maximum de 9 ans. L'analyse des cas concernés montre que ces ruptures peuvent correspondre à des périodes sans observation IDMC pour le pays considéré. Elles sont donc conservées et ne sont pas imputées comme des données manquantes.

## 6.  Qualité de données

In [43]:
### Vérifier la qualité de total IDMC

## Cette variable servira à construire la target

idmc = dfs["idmc"].copy()

total = pd.to_numeric(
    idmc["total"],
    errors="coerce"
)

print("Missing :", total.isna().sum())
print("Zéro :", (total == 0).sum())
print("Négatifs :", (total < 0).sum())
print("Minimum :", total.min())
print("Maximum :", total.max())
print("Médiane :", total.median())

Missing : 0
Zéro : 0
Négatifs : 0
Minimum : 2
Maximum : 28000000
Médiane : 204000.0


In [44]:
### Vérification des outliers sans les supprimer

display(
    total.describe(
        percentiles=[
            .01, .05, .25,
            .50, .75, .95, .99
        ]
    )
)

count    9.340000e+02
mean     1.272705e+06
std      3.560060e+06
min      2.000000e+00
1%       1.100000e+02
5%       9.000000e+02
25%      2.500000e+04
50%      2.040000e+05
75%      6.805000e+05
95%      5.614500e+06
99%      2.403700e+07
max      2.800000e+07
Name: total, dtype: float64

In [45]:
idmc_ml = (
    idmc
    .groupby(
        ["year", "coo_id", "coo_iso"],
        as_index=False,
        dropna=False
    )
    .agg(
        idmc_total=("total", "sum")
    )
)

In [46]:
print(
    "Doublons année × pays :",
    idmc_ml.duplicated(
        ["year", "coo_id"]
    ).sum()
)

Doublons année × pays : 0


### Contrôler la future target >30 %

- permet de clacul d'une croissance historique


*Combien d'IDPs ce pays possède-t-il cette année ?* - *« À partir de l'année T, ce pays connaîtra-t-il une hausse supérieure à 30 % à T+1 ? »*

In [48]:
# Trier chronologiquement chaque pays
idmc_ml = (
    idmc_ml
    .sort_values(["coo_id", "year"])
)

# Chercher l'année suivante
idmc_ml["year_t1"] = (
    idmc_ml
    .groupby("coo_id")["year"]
    .shift(-1)
)

# Chercher le stock IDMC de l'observation suivante
idmc_ml["total_t1"] = (
    idmc_ml
    .groupby("coo_id")["idmc_total"]
    .shift(-1)
)


# Vérifier que T+1 est réellement l'année suivante
idmc_ml["valid_t1"] = (
    idmc_ml["year_t1"]
    == idmc_ml["year"] + 1
)

# Définir les lignes sur lesquelles la croissance peut être calculée
valid = (
    idmc_ml["valid_t1"]
    & (idmc_ml["idmc_total"] > 0)
)

# Calculer la croissance entre T et T+1
idmc_ml.loc[valid, "growth_t1"] = (
    (
        idmc_ml.loc[valid, "total_t1"]
        - idmc_ml.loc[valid, "idmc_total"]
    )
    / idmc_ml.loc[valid, "idmc_total"]
)


# Construire la cible ML
idmc_ml.loc[
    valid,
    "hausse_critique_suivante"
] = (
    idmc_ml.loc[valid, "growth_t1"] > 0.15
).astype(int)


# Identifier les stocks IDMC égaux à zéro
zero_base = (
    idmc_ml["idmc_total"] == 0
)

print(
    "Bases IDMC nulles :",
    zero_base.sum()
)


# Permet de connaitre un phénomène de déplacement apparaît là où aucun stock n'était précédemment observé.
idmc_ml["new_emergence"] = (
    (idmc_ml["idmc_total"] == 0)
    & (idmc_ml["total_t1"] > 0)
).astype(int)


# le nombre d'observations de chaque classe
display(
    idmc_ml[
        "hausse_critique_suivante"
    ]
    .value_counts(dropna=False)
)


# Calculer les proportions
display(
    idmc_ml[
        "hausse_critique_suivante"
    ]
    .value_counts(
        normalize=True,
        dropna=False
    ) * 100
)

Bases IDMC nulles : 0


hausse_critique_suivante
0.0    654
1.0    181
NaN     99
Name: count, dtype: int64

hausse_critique_suivante
0.0    70.021413
1.0    19.379015
NaN    10.599572
Name: proportion, dtype: float64

In [49]:
idmc_ml = (
    idmc_ml
    .sort_values(["coo_id", "year"])
)

display(idmc_ml)

,year,coo_id,coo_iso,idmc_total,year_t1,total_t1,valid_t1,growth_t1,hausse_critique_suivante,new_emergence
19,2009,2,AFG,297000,2010.0,352000.0,True,0.185185,1.0,0
65,2010,2,AFG,352000,2011.0,450000.0,True,0.278409,1.0,0
112,2011,2,AFG,450000,2012.0,492000.0,True,0.093333,0.0,0
161,2012,2,AFG,492000,2013.0,631000.0,True,0.282520,1.0,0
210,2013,2,AFG,631000,2014.0,805000.0,True,0.275753,1.0,0
...,...,...,...,...,...,...,...,...,...,...
701,2021,263,AB9,15000,2022.0,56000.0,True,2.733333,1.0,0
761,2022,263,AB9,56000,2023.0,42000.0,True,-0.250000,0.0,0
820,2023,263,AB9,42000,2024.0,42000.0,True,0.000000,0.0,0
880,2024,263,AB9,42000,2025.0,58000.0,True,0.380952,1.0,0
